### PR TITLE
Add statistics rules for aggregation and semi-join

### PR DESCRIPTION
**Pull request by findepi**

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/cost/TestIntersectStatsRule.java [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+                        .symbolStats("o1", assertion -> assertion
+                                .lowValue(5)
+                                .highValue(10)
+                                .dataSizeUnknown()
+                                .distinctValuesCount(4)
+                                .nullsFraction(0.2)));
+    }
+
+    @Test
+    public void testIntersectWhenRangesAreSeparated()
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/test/java/com/facebook/presto/cost/TestIntersectStatsRule.java]: We compute 1 instead of 0 because 1 is the minimal row count result of `AggregationStatsRule#groupBy`.
We could fix this.. but what if we calculate something incorrectly and 0 row count turns into some damage later on?
just thinking..

@kokosing do you want to add something?

File: presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java [hunk 2]
```diff
@@ -75,21 +75,50 @@ public void testJoin()
...
+    @Test
+    public void testIntersect()
+    {
         statisticsAssertion.check("SELECT * FROM nation INTERSECT SELECT * FROM nation",
-                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(0, 1.)));
+                // real count is 25, estimation cannot know all rows are duplicate.
+                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(.7, .9)));
+
+        statisticsAssertion.check("SELECT * FROM orders WHERE o_custkey < 900 INTERSECT SELECT * FROM orders WHERE o_custkey > 600",
+                // TODO fix INTERSECT stats calculation as custkey values distirbution is pretty linear
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java]: @kokosing could you?

File: presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java [hunk 3]
```diff
@@ -75,21 +75,50 @@ public void testJoin()
...
+                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(4, 5)));
+    }
 
+    @Test
+    public void testExcept()
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java]: There is a `ImplementIntersectAndExceptAsUnion` rule

File: presto-main/src/main/java/com/facebook/presto/cost/SemiJoinStatsRule.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        return PATTERN;
+    }
+
+    @Override
+    public Optional<PlanNodeStatsEstimate> calculate(PlanNode node, StatsProvider statsProvider, Lookup lookup, Session session, Map<Symbol, Type> types)
+    {
+        SemiJoinNode semiJoinNode = (SemiJoinNode) node;
+        PlanNodeStatsEstimate sourceStats = statsProvider.getStats(semiJoinNode.getSource());
+
+        // For now we just propagate statistics for source symbols.
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/SemiJoinStatsRule.java]: Yes, this is quite accurate. Semi join has two sources:
- `source`
- `filteringSource` (which isn't used for filtering, just as you described)

Stats for boolean column would be useful, because our stats model currently doesn't allow expressing things like "number of trues", "number of falses". The only thing we can calculate for this new porjected boolean columns would be NDV=2.

File: presto-main/src/main/java/com/facebook/presto/cost/SemiJoinStatsRule.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        // Handling semiJoinOutput symbols requires support for correlation for boolean columns.
+
+        // TODO add UNKNOWN_STATS for ((SemiJoinNode) node).getSourceJoinSymbol(), drop unused symbols... there should be a test for that, right?
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/SemiJoinStatsRule.java]: It's outdated.

- adding UNKNOWN_STATS isn't required, since
      `com.facebook.presto.cost.PlanNodeStatsEstimate#getSymbolStatistics`
      defaults to UNKNOWN_STATS anyway

 - dropping unused symbols doesn't make sense for SemiJoinNode, since the
      node currently always copies symbols from input to output

**Comment by findepi** [comment]
AC

File: presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java]: I wish I could remember this...
I suspect it is because INTERSECT is replaced with an aggregation and then once it happens it is no longer trivial to estimate stats.

File: presto-main/src/test/java/com/facebook/presto/cost/TestIntersectStatsRule.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rschlussel2) [reply presto-main/src/test/java/com/facebook/presto/cost/TestIntersectStatsRule.java]: Why have an IntersectStatsRule at all then?  Maybe we should get rid of it if it's not going to be used.

[the remark this answers is not in the document]
Comment (kokosing) [reply presto-main/src/test/java/com/facebook/presto/cost/TestIntersectStatsRule.java]: I do not remember why it was created. Maybe we was not aware of this rewrite. Or maybe we had too much processing power in Teradata ;)

From the other point of view. If we have then what is the benefit of its removal? I could image a CBO rule that could leverage this at some point in the future. For example, I believe there are multiple ways of how Intersect could be implemented and deciding which way should be used could be a cost-based decision.

File: presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java [hunk 6]
```diff
@@ -266,26 +266,33 @@ private static PlanNodeStatsEstimate filterStatsForUnknownExpression(PlanNodeSta
...
-                return visitExpression(node, context);
+                return Optional.empty();
             }
 
-            Symbol inValueSymbol = Symbol.from(node.getValue());
-            SymbolStatsEstimate symbolStats = input.getSymbolStatistics(inValueSymbol);
-            double notNullValuesBeforeIn = input.getOutputRowCount() * (1 - symbolStats.getNullsFraction());
+            Optional<Symbol> inValueSymbol = asSymbol(node.getValue());
+            SymbolStatsEstimate inValueStats = getExpressionStats(node.getValue());
+            if (Objects.equals(inValueStats, UNKNOWN_STATS)) {
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java]: `UNKNOWN_STATS.equals(inValueStats)`

File: presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java [hunk 7]
```diff
@@ -266,26 +266,33 @@ private static PlanNodeStatsEstimate filterStatsForUnknownExpression(PlanNodeSta
...
 
             if (isNaN(statsSum.getOutputRowCount())) {
-                return visitExpression(node, context);
+                return Optional.empty();
             }
 
-            Symbol inValueSymbol = Symbol.from(node.getValue());
-            SymbolStatsEstimate symbolStats = input.getSymbolStatistics(inValueSymbol);
-            double notNullValuesBeforeIn = input.getOutputRowCount() * (1 - symbolStats.getNullsFraction());
+            Optional<Symbol> inValueSymbol = asSymbol(node.getValue());
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java]: This looks like an obfuscation. Why don't just do

```
if(node.getValue() instanceof SymbolReference){
 ....
} else {
 ....
}
```

File: presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java [hunk 8]
```diff
@@ -266,26 +266,33 @@ private static PlanNodeStatsEstimate filterStatsForUnknownExpression(PlanNodeSta
...
+            double notNullValuesBeforeIn = input.getOutputRowCount() * (1 - inValueStats.getNullsFraction());
 
-            return Optional.of(input.mapOutputRowCount(rowCount -> min(statsSum.getOutputRowCount(), notNullValuesBeforeIn))
-                    .mapSymbolColumnStatistics(inValueSymbol, oldSymbolStats -> newSymbolStats));
+            PlanNodeStatsEstimate estimate = input.mapOutputRowCount(rowCount -> min(statsSum.getOutputRowCount(), notNullValuesBeforeIn));
+
+            if (inValueSymbol.isPresent()) {
+                SymbolStatsEstimate newSymbolStats = statsSum.getSymbolStatistics(inValueSymbol.get())
+                        .mapDistinctValuesCount(newDistinctValuesCount -> min(newDistinctValuesCount, inValueStats.getDistinctValuesCount()));
+                estimate = estimate.mapSymbolColumnStatistics(inValueSymbol.get(), oldSymbolStats -> newSymbolStats);
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java]: I would just build it with a builder instead

File: presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java [hunk 9]
```diff
@@ -99,6 +99,15 @@ public Checks verifyCharacterColumnStatistics(String columnName, MetricCompariso
             return this;
         }
 
+        public Checks verifyNoColumnStatistics(String columnName)
+        {
+            noEstimate(nullsFraction(columnName));
```
Comment (arhimondr) [inline presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java]: Some of the methods in this class start with `verify` some of them don't.  This bit is confusing. Maybe `expectNoEstimate`? or `verifyNoEstimate`? Although it should rather be a part of a separate PR.

File: presto-main/src/test/java/com/facebook/presto/cost/TestIntersectStatsRule.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/test/java/com/facebook/presto/cost/TestIntersectStatsRule.java]: I agree, if this is not needed now, we should remove it. We were generally cautious not to do things within CBO that are already handled outside.

File: presto-main/src/main/java/com/facebook/presto/cost/AggregationStatsRule.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+        List<Symbol> groupBySymbols = getOnlyElement(aggregationNode.getGroupingSets());
+
+        PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.builder();
+        for (Symbol groupBySymbol : groupBySymbols) {
+            SymbolStatsEstimate symbolStatistics = sourceStats.getSymbolStatistics(groupBySymbol);
+            result.addSymbolStatistics(groupBySymbol, symbolStatistics.mapNullsFraction(nullsFraction -> {
+                if (isPositiveOrNan(nullsFraction)) {
+                    double distinctValuesCount = symbolStatistics.getDistinctValuesCount();
+                    return 1.0 / (distinctValuesCount + 1);
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/AggregationStatsRule.java]: >  or `(1.0 / (distinctValuesCount + 1)) * aggregtaion_output_estimated_rows_count` might be better.

I guess your formula has a typo -- since we calculate nulls _fraction_, it should go without `* aggregtaion_output_estimated_rows_count` part, right? But then, isn't it what's written here? Please clarify.

File: presto-main/src/main/java/com/facebook/presto/cost/SemiJoinStatsCalculator.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+            Symbol filteringSourceJoinSymbol,
+            BiFunction<SymbolStatsEstimate, SymbolStatsEstimate, Double> retainedNdvProvider)
+    {
+        SymbolStatsEstimate sourceJoinSymbolStats = sourceStats.getSymbolStatistics(sourceJoinSymbol);
+        SymbolStatsEstimate filteringSourceJoinSymbolStats = filteringSourceStats.getSymbolStatistics(filteringSourceJoinSymbol);
+
+        double retainedNDV = retainedNdvProvider.apply(sourceJoinSymbolStats, filteringSourceJoinSymbolStats);
+        double filterFactor = sourceJoinSymbolStats.getValuesFraction() * retainedNDV / sourceJoinSymbolStats.getDistinctValuesCount();
+
+        SymbolStatsEstimate newSourceJoinSymbolStats = SymbolStatsEstimate.buildFrom(sourceJoinSymbolStats)
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/SemiJoinStatsCalculator.java]: i retained original wording

File: presto-main/src/main/java/com/facebook/presto/cost/SemiJoinStatsCalculator.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+                        max(sourceJoinSymbolStats.getDistinctValuesCount() * MAX_ANTI_JOIN_FILTER_COEFFICIENT,
+                                sourceJoinSymbolStats.getDistinctValuesCount() - filteringSourceJoinSymbolStats.getDistinctValuesCount()));
+    }
+
+    private PlanNodeStatsEstimate compute(
+            PlanNodeStatsEstimate sourceStats,
+            PlanNodeStatsEstimate filteringSourceStats,
+            Symbol sourceJoinSymbol,
+            Symbol filteringSourceJoinSymbol,
+            BiFunction<SymbolStatsEstimate, SymbolStatsEstimate, Double> retainedNdvProvider)
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/SemiJoinStatsCalculator.java]: This is called in two places and the callbacks' formulae are different

File: presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java (hunk 7 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java]: I have different opinion on this. Maybe because `instanceof`s scare me :)

File: presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java (hunk 9 above)
Comment (findepi) [reply presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java]: Hm... I somewhat agree. However, let's put this cleanup aside for now, since this PR doesn't intro any new mess here.

File: presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java (hunk 8 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java]: I left the existing wording.

File: presto-main/src/main/java/com/facebook/presto/cost/AggregationStatsRule.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/AggregationStatsRule.java]: Let me put here a more detailed explanation.

When doing a group by by a single column, `null` values are folded into a single `null`. So the `1 / (ndv + 1)` is perfect.

However when grouping by multiple columns, values are folded on row basis.

```
SELECT *
FROM (values (null, 1), (null, 2), (null, 3), (null, 4), ... ,(null, N)) as t(x,y)
GROUP BY (x, y)
```
results in 
```
null | 1
null | 2
null | 3
null | 4
...
null | N
```

In this case a `null_fraction * output_rows` is a better estimate than a `1 / (ndv + 1)`. Not ideal though, as `null` might correlate with other columns. But because we don't know anything about correlation, the `output_rows` estimate is also off.

File: presto-main/src/main/java/com/facebook/presto/cost/SemiJoinStatsCalculator.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/SemiJoinStatsCalculator.java]:
```
public static PlanNodeStatsEstimate computeAntiJoin(PlanNodeStatsEstimate sourceStats, PlanNodeStatsEstimate filteringSourceStats, Symbol sourceJoinSymbol, Symbol filteringSourceJoinSymbol)
{
        SymbolStatsEstimate sourceJoinSymbolStats = sourceStats.getSymbolStatistics(sourceJoinSymbol);
        SymbolStatsEstimate filteringSourceJoinSymbolStats = filteringSourceStats.getSymbolStatistics(filteringSourceJoinSymbol);
        double retainedNdv = max(sourceJoinSymbolStats.getDistinctValuesCount() * MIN_ANTI_JOIN_FILTER_COEFFICIENT,
                                sourceJoinSymbolStats.getDistinctValuesCount() - filteringSourceJoinSymbolStats.getDistinctValuesCount())
        return compute(sourceStats, filteringSourceStats, sourceJoinSymbol, filteringSourceJoinSymbol, retainedNdv);
}

```

[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/SemiJoinStatsCalculator.java]: I find it much more readable. And since you no longer need

```
SymbolStatsEstimate filteringSourceJoinSymbolStats = filteringSourceStats.getSymbolStatistics(filteringSourceJoinSymbol);
```

inside the `compute` method, as long as `filteringSourceJoinSymbol`, number of non space symbols in this file shouldn't be any bigger =)

File: presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java (hunk 6 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java]: Since this is a nit, feel free to leave it as is. But, please, just sate my curiosity. How's is the `Objects.equals(inValueStats, UNKNOWN_STATS)` style any better than a simple `equals`? Neither the constant, nor the `inValueStats` could be `null`. So the `Objects.equals` seems to be misleading. Since it implies that it's parameters may be `null`.

File: presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java (hunk 7 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java]: You can always introduce an utility method, `isSymbolReference`. But i think we just do `instanceof` in other parts of our code.

My point basically that there are simple cases. When your expression is a simple symbol reference, and when the expression is a complex expression. But it took me a while to figure this out.

File: presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java (hunk 8 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java]: No change is required. Feel free to leave it as is.

I just looks like a yet another lambda abuse.

File: presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java (hunk 6 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java]: I prefer `Objects.equals(a,b)` over `a.equals(b)` because the latter is asymmetric

File: presto-main/src/main/java/com/facebook/presto/cost/AggregationStatsRule.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/AggregationStatsRule.java]: UPD:

Based on the offline discussion. What we are doing right now seems to be the best.

File: presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java (hunk 7 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java]: You might be right. This sounds a bit as a matter of taste. And I'm reluctant to do any dispensable changes in this particular class at this point, as I don't want to accidentally break something resolving conflicts during next rebase.

File: presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java (hunk 8 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java]: i sometimes had an impression that we shouldn't have these `map*` (`mapSymbolColumnStatistics`, `mapDistinctValuesCount`, ..) methods at all. We should approach this holistically though -- review use cases, see if it would be better not to have these methods....  However -- slightly later, to avoid conflicts.
